### ISSUE TITLE
Adopt OpenAPI Specification for describing the DSpace REST API

### ISSUE THREAD

**Issue by saschaszott** [DSpace/DSpace#10118]
## Description of Improvement Request

The current DSpace REST API is described informally using a free-text document (RestContract, https://github.com/DSpace/RestContract). While this approach provides a basic overview of available endpoints, it lacks a formal (machine-readable), standardized structure. 

This proposal suggests formally documenting the DSpace REST API using the **OpenAPI Specification** (https://www.openapis.org/). This will enhance usability, consistency, and integration capabilities for developers and external systems.

Formalizing the API using OpenAPI Specification comes with several benefits, e.g.,

- Enforcing a standard structure for API definitions
- Providing auto-generated, always-up-to-date documentation
- Supporting client library generation in multiple languages

**Comment by tdonohue** [comment]
While I'm definitely in favor of some sort of automated description for the DSpace REST API, we had previously discussed using ALPS simply because it's supported by Spring Data REST (which is what we model our API against): https://docs.spring.io/spring-data/rest/reference/metadata.html   See also #2863 

So, we would need to likely choose between this ticket and #2863 .  In any case, we've yet to find a volunteer to implement ALPS support

**Comment by mirand863** [comment]
I have to add OpenAPI to DSpace v9.2 for my network admin to configure bot/DDoS protection. Would it be interesting if I create a PR when I finish?

**Comment by mirand863** [comment]
@tdonohue and @saschaszott I created PR https://github.com/DSpace/DSpace/pull/11720 for this issue. I believe the tests that failed were due to connectivity issues, but I do not have permission to rerun the jobs. Can you please rerun the failed jobs?

**Comment by mirand863** [comment]
@tdonohue and @saschaszott 

I have updated the GitHub CI/CD to create the folder /tmp/docker since a command was failing when this folder did not exist. Now all the tests have passed. Can you please review #11720 when you have some time? Or how can I request a reviewer?

**Comment by tinsch** [comment]
This might be a long shot, but since we currently use the HAL Browser, it could be worth investigating the use of the HAL explorer. It already includes an easy [way to add documentation](https://toedter.github.io/hal-explorer/snapshot/reference-doc/#_curie_based_documentation), and also the HAL browser seems to be not actively maintained anymore.

With this approach we wouldn't need to add another layer of documentation and could stay close to the current implementation.

**Comment by tdonohue** [comment]
@tinsch : Yes, we already have a ticket for moving to the HAL explorer: https://github.com/DSpace/DSpace/issues/3017  Unfortunately, though, it's not a simple migration because the HAL explorer doesn't support all the features of the HAL Browser, e.g. authentication.  That was a concern in the past, but it may be that we need to reanalyze if it's worth moving the the HAL explorer anyways.

In any case, I'm still open to moving to the HAL explorer.  But doing so means we will lose some features that are only available in the HAL Browser (but, in exchange for some features only available in the HAL explorer).  

If someone is interested in trying out a migration to the HAL explorer, we could always reopen this discussion in a future Developers' meeting.
